File: Combined/DesktopApp/src/featurecalculator.cpp
﻿#include "featurecalculator.h"

FeatureCalculator::FeatureCalculator()
{
    FeatureCalculator::featTableName = configHandler->getTableName("features");
}

void FeatureCalculator::run()
{

    dbHandler->addTableToDb(configHandler->getColumnsInfoPath(), "features_list", featTableName);

    pqxx::result result = dbHandler->executeQuery(dbHandler->connection, "SELECT * FROM " + configHandler->getTableName("metadata") + ";");

    int count = 0;
    for (int rownum=0; rownum < result.size(); ++rownum)
    {
        // Read in the image
        const pqxx::row row = result[rownum];
        const char * filePath = row["file_path"].c_str();
        count++;

        DicomImage * dcmImage = new DicomImage(filePath);
        uint16_t * pixelData = (uint16_t *) (dcmImage->getOutputData(16)); // This scales the pixel values so that the intensity range is 0 - 2^16 instead of 0 - 2^bits_stored

        // Convert the image from uint16 to double
        uint64_t height = dcmImage->getHeight();
        uint64_t width = dcmImage->getWidth();

        imageUnsigned = cv::Mat(height, width, CV_16U, pixelData);
        imageUnsigned.convertTo(imageFloat, CV_64F);

        // Get the bits_stored for scaling intensities back down from 16 bits to their bits_stored value
        DcmFileFormat file_format;
        file_format.loadFile(filePath);

        DcmTagKey bitsStoredKey(40, 257); // bits stored
        OFString bitsStoredValue;
        file_format.getDataset()->findAndGetOFString(bitsStoredKey, bitsStoredValue);

        if (bitsStoredValue == "12") {
            imageFloat = imageFloat / 16.0; // 16 - 12 = 4 bits = 2^4 = 16
        } else if (bitsStoredValue == "15") {
            imageFloat = imageFloat / 2.0; // 16 - 15 = 1 bit = 2^1 = 2
        } else if (bitsStoredValue == "10") {
            imageFloat = imageFloat / 64.0; // 16 - 10 = 6 bits = 2^6 = 64
        } else if (bitsStoredValue == "14") {
            imageFloat = imageFloat / 2.0; // 16 - 14 = 2 bits = 2^2 = 4
        } else {
            // log "BITS STORED NOT EXPECTED"
        }

        imageFloat.convertTo(imageUnsigned, CV_16U, 1, -0.5); // This rounding might cause inconsistencies

        imageFloat = preprocessing(atoi(bitsStoredValue.c_str()));

        cv::resize(imageFloat, imageResize, cv::Size(200, 200), 0, 0, cv::INTER_AREA);

        horProfile = calcHorProf();
        vertProfile = calcVertProf();

        store(filePath, horProfile, vertProfile);

        delete dcmImage;
    }
}

void FeatureCalculator::store(std::string filePath, cv::Mat horProfile, cv::Mat vertProfile)
{
    // Create SQL query
    std::vector<double> horVec(horProfile.begin<double>(), horProfile.end<double>());
    std::vector<double> vertVec(vertProfile.begin<double>(), vertProfile.end<double>());

    std::vector<std::string> horVecString(horVec.size());
    std::transform(horVec.begin(), horVec.end(), horVecString.begin(), [](const double& val){return std::to_string(val);});

    std::vector<std::string> vertVecString(vertVec.size());
    std::transform(vertVec.begin(), vertVec.end(), vertVecString.begin(), [](const double& val){return std::to_string(val);});

    std::string sqlQuery = "INSERT INTO " + featTableName + " (file_name, file_path, hor_profile, vert_profile) VALUES ('" +
            filePath.substr(filePath.find_last_of("/") + 1) + "', '" + filePath + "', '{" + boost::algorithm::join(horVecString, ", ") +
            "}', '{" + boost::algorithm::join(vertVecString, ", ") + "}');";

    dbHandler->executeQuery(dbHandler->connection, sqlQuery);
}

cv::Mat FeatureCalculator::calcHorProf()
{
    cv::Mat horProfile(1, 200, CV_64F);
//    cv::cuda::reduce(imageResize, horProfile, 0, cv::REDUCE_AVG, CV_64F);
    cv::reduce(imageResize, horProfile, 1, cv::REDUCE_AVG, CV_64F);

    return horProfile;
}

cv::Mat FeatureCalculator::calcVertProf()
{
    cv::Mat vertProfile(200, 1, CV_64F);
//    cv::cuda::reduce(imageResize, vertProfile, 1, cv::REDUCE_AVG, CV_64F);
    cv::reduce(imageResize, vertProfile, 1, cv::REDUCE_AVG, CV_64F);

    return vertProfile;
}

cv::Mat FeatureCalculator::preprocessing(uint8_t bitsStored)
{
    // Normalize image by dividing the intensities by the highest possible intensity so that it's dynamic range is between 0.0 and 1.0
    float highestPossibleIntensity = pow(2, bitsStored) - 1;
    imageUnsigned.convertTo(imageFloat, CV_64F);
    imageFloat = imageFloat / highestPossibleIntensity;

    // Calculate the 1st and 99th percentile
    imageFloatFlat = imageFloat.reshape(1, 1);
    cv::Mat imageNormSorted(imageFloatFlat.size(), CV_64F);
    cv::sort(imageFloatFlat, imageNormSorted, cv::SORT_ASCENDING);

    uint64_t nPixels = imageUnsigned.rows * imageUnsigned.cols;

    uint64_t firstIndex = 0.01 * nPixels;
    uint64_t ninenineIndex = 0.99 * nPixels;

    double firstPercentile = imageNormSorted.at<double>(firstIndex);
    double nineninePercentile = imageNormSorted.at<double>(ninenineIndex);

    // Perform the contrast stretch
    cv::Mat enhancedImage = (imageFloat - firstPercentile) / (nineninePercentile - firstPercentile) * 1.0;
    enhancedImage.setTo(0.0, imageFloat < firstPercentile);
    enhancedImage.setTo(1.0, imageFloat > nineninePercentile);

    // Calculate the median
    imageFloatFlat = enhancedImage.reshape(1, 1);
    std::vector<double> vecFromMat;
    imageFloatFlat.copyTo(vecFromMat);
    std::nth_element(vecFromMat.begin(), vecFromMat.begin() + vecFromMat.size() * 0.5, vecFromMat.end());
    double median = vecFromMat[vecFromMat.size() / 2];

    // Threshold the image at median
    enhancedImage.convertTo(imageFloat, CV_64F);
    cv::Mat imageBinarized(imageUnsigned.rows, imageUnsigned.cols, CV_64F);
    cv::threshold(imageFloat, imageBinarized, median, 1.0, cv::THRESH_BINARY);

    // Crop image
    cv::Mat points;
    cv::findNonZero(imageBinarized, points);
    cv::Rect bb = cv::boundingRect(points);

    cv::Mat imageCropped;
    if (!points.empty()) {
        imageCropped = imageFloat(bb);
    } else {
        imageCropped = imageFloat;
    }

    // Scale image
    double scalePercent = 0.5;
    unsigned width = imageCropped.cols * scalePercent;
    unsigned height = imageCropped.rows * scalePercent;

    cv::Mat imageDownsize(height, width, CV_64F);

    cv::resize(imageCropped, imageDownsize, cv::Size(width, height), 0.5, 0.5, cv::INTER_AREA);

    return imageDownsize;
}


FeatureCalculator * FeatureCalculator_new() { 
    return new FeatureCalculator(); 
}

void FeatureCalculator_run(FeatureCalculator * featCalculator) {
    featCalculator->run();
}

File: Go/CxrClassify/featureCalculator/featureCalculator.go
package featureCalculator

import (
	"CxrClassify/configHandler"
	"CxrClassify/databaseHandler"
	"CxrClassify/runnable"
	"fmt"
	"image"
	"image/png"
	"log"
	"math"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"time"

	"github.com/suyashkumar/dicom"
	"github.com/suyashkumar/dicom/pkg/tag"
	"gocv.io/x/gocv"
)

type FeatureCalculator struct {
	runnable.Runnable

	featTableName string

	_ func() `constructor:"init"`
}

func (f *FeatureCalculator) init() {

}

func (f *FeatureCalculator) Setup(configHandler *configHandler.ConfigHandler, databaseHandler *databaseHandler.DatabaseHandler) {
	f.featTableName = configHandler.GetTableName("features")
	f.SetupRunnable(configHandler, databaseHandler)
}

func (f FeatureCalculator) Run() {
	k, err := os.OpenFile("testlogfile", os.O_RDWR|os.O_CREATE|os.O_APPEND, 0666)
	if err != nil {
		log.Fatalf("error opening file: %v", err)
	}
	defer k.Close()
	log.SetOutput(k)

	log.Printf("Calculating features from images")

	f.AttemptUpdateText("Calculating features")
	f.AttemptUpdateProBarBounds(0, f.Expected_num_files)

	f.DatabaseHandler.AddTableToDb(f.ConfigHandler.GetColumnsInfoName(), "features_list", f.featTableName)

	f.AttemptUpdateProBarValue(0)
	sqlQuery := "SELECT file_path FROM " + f.ConfigHandler.GetTableName("metadata") + ";"
	rows, _ := f.DatabaseHandler.ExecuteQuery(f.DatabaseHandler.Connection, sqlQuery)

	var filePath string
	if err == nil {
		count := 0
		for rows.Next() {
			log.Printf("Calculating for image number: %d File: %s", count+1, filePath)
			err = rows.Scan(&filePath)
			if err == nil {
				dataset, err := dicom.ParseFile(filePath, nil)
				if err != nil {
					log.Println(err)
				}

				pixelDataElement, err := dataset.FindElementByTag(tag.PixelData)
				if err != nil {
					log.Println(err)
				}
				pixelDataInfo := dicom.MustGetPixelDataInfo(pixelDataElement.Value)
				frame := pixelDataInfo.Frames[0]

				pixelData, err := frame.GetImage() // Intensity range of 0 - 2^bits_stored

				if err != nil {
					log.Println(err)
				}

				// bounds := pixelData.Bounds()
				// width, height := bounds.Max.X, bounds.Max.Y

				fileObj, err := os.Create(fmt.Sprintf("tmp.png"))
				if err != nil {
					log.Println(err)
				}
				err = png.Encode(fileObj, pixelData)
				if err != nil {
					log.Println(err)
				}
				err = fileObj.Close()
				if err != nil {
					log.Println(err)
				}

				time.Sleep(time.Second)

				imageUnsigned := gocv.IMRead("tmp.png", gocv.IMReadAnyDepth)

				pixelDataElement, err = dataset.FindElementByTag(tag.Tag{Group: 40, Element: 257})
				if err != nil {
					log.Println(err)
				}
				value := pixelDataElement.Value.String()
				value = value[1 : len(value)-1]
				bitsStored, err := strconv.ParseFloat(value, 64)
				if err != nil {
					log.Println(err)
				}

				highestPossibleIntensity := math.Pow(2, bitsStored) - 1

				imageFloat := gocv.NewMat()
				defer imageFloat.Close()
				imageUnsigned.ConvertTo(&imageFloat, gocv.MatTypeCV32F)

				imageFloat.DivideFloat(float32(highestPossibleIntensity)) // Conversion loses precision

				imageFloatFlat := imageFloat.Reshape(1, 1)
				imageNormSorted := gocv.NewMatWithSize(imageFloatFlat.Rows(), imageFloatFlat.Cols(), gocv.MatTypeCV32F)
				gocv.Sort(imageFloatFlat, &imageNormSorted, gocv.SortAscending)

				nPixels := float32(imageUnsigned.Rows() * imageUnsigned.Cols())

				firstIndex := int(0.01 * nPixels)
				ninenineIndex := int(0.99 * nPixels)

				firstPercentile := imageNormSorted.GetFloatAt(0, firstIndex)
				nineninePercentile := imageNormSorted.GetFloatAt(0, ninenineIndex)

				// Perform the contrast stretch
				enhancedImage := imageFloat.Clone()
				enhancedImage.SubtractFloat(firstPercentile)
				enhancedImage.DivideFloat(nineninePercentile - firstPercentile)

				for i := 0; i < enhancedImage.Rows(); i++ {
					for j := 0; j < enhancedImage.Cols(); j++ {

						if enhancedImage.GetFloatAt(i, j) < firstPercentile {
							// log.Printf("Float: %f", enhancedImage.GetFloatAt(i, j))
							enhancedImage.SetFloatAt(i, j, 0.0)
						} else if enhancedImage.GetFloatAt(i, j) > nineninePercentile {
							// log.Printf("Float: %f", enhancedImage.GetFloatAt(i, j))
							enhancedImage.SetFloatAt(i, j, 1.0)
						}

					}
				}

				// Calculate the median
				enhancedImageFlat := enhancedImage.Reshape(1, 1)
				enhancedImageSort := gocv.NewMatWithSize(enhancedImageFlat.Rows(), enhancedImageFlat.Cols(), gocv.MatTypeCV32F)
				gocv.Sort(enhancedImageFlat, &enhancedImageSort, gocv.SortAscending)
				medianIndex := int(0.50 * nPixels)
				median := enhancedImageSort.GetFloatAt(0, medianIndex)

				// min, max, _, _ := gocv.MinMaxLoc(enhancedImageFlat)

				// log.Printf("Min: %f, Max: %f", min, max)
				// log.Println(median)

				// Threshold the image at median
				enhancedImageDouble := gocv.NewMat()
				defer enhancedImageDouble.Close()
				enhancedImage.ConvertTo(&enhancedImageDouble, gocv.MatTypeCV32F)
				imageBinarized := gocv.NewMatWithSize(enhancedImage.Rows(), enhancedImage.Cols(), gocv.MatTypeCV32F)
				gocv.Threshold(enhancedImageDouble, &imageBinarized, median, 1.0, gocv.ThresholdBinary)

				// Crop image
				points := gocv.NewMat()
				gocv.FindNonZero(imageBinarized, &points)
				pointVec := gocv.NewPointVectorFromMat(points)
				bb := gocv.BoundingRect(pointVec)

				imageCropped := gocv.NewMat()
				if pointVec.Size() > 0 {
					imageCropped = enhancedImageDouble.Region(bb)
				} else {
					log.Println("ERROR")
				}

				// window := gocv.NewWindow("Hello")
				// window.IMShow(imageCropped)
				// window.WaitKey(0)

				// Scale image
				scalePercent := 0.5
				newWidth := int(float64(imageCropped.Cols()) * scalePercent)
				newHeight := int(float64(imageCropped.Rows()) * scalePercent)

				imageDownsize := gocv.NewMatWithSize(newHeight, newWidth, gocv.MatTypeCV64F)
				gocv.Resize(imageCropped, &imageDownsize, image.Pt(newWidth, newHeight), 0.5, 0.5, gocv.InterpolationArea)

				imageResize := gocv.NewMatWithSize(200, 200, gocv.MatTypeCV64F)
				gocv.Resize(imageDownsize, &imageResize, image.Pt(200, 200), 0, 0, gocv.InterpolationArea)

				horProfile := gocv.NewMatWithSize(1, 200, gocv.MatTypeCV64F)
				gocv.Reduce(imageResize, &horProfile, 0, gocv.ReduceAvg, gocv.MatTypeCV64F)

				vertProfile := gocv.NewMatWithSize(200, 1, gocv.MatTypeCV64F)
				gocv.Reduce(imageResize, &vertProfile, 1, gocv.ReduceAvg, gocv.MatTypeCV64F)

				// Create SQL Query

				log.Printf("Storing the calculated features into the database.")

				horVecString := []string{}
				vertVecString := []string{}

				for i := 0; i < 200; i++ {
					horVecString = append(horVecString, fmt.Sprint(enhancedImage.GetFloatAt(0, i)))
					vertVecString = append(vertVecString, fmt.Sprint(enhancedImage.GetFloatAt(i, 0)))
				}

				sqlQuery := "INSERT INTO " + f.ConfigHandler.GetTableName("features") + " (file_name, file_path, hor_profile, vert_profile) VALUES ('" + filepath.Base(filePath) + "', '" + filePath + "', '{" + strings.Join(horVecString, ", ") + "}', '{" + strings.Join(vertVecString, ", ") + "}');"

				connection := f.DatabaseHandler.OpenConnection(false)

				f.DatabaseHandler.ExecuteQuery(connection, sqlQuery)

				connection.Close()

				// cv::Mat imageCropped;
				// if (!points.empty()) {
				// 	imageCropped = imageFloat(bb);
				// } else {
				// 	imageCropped = imageFloat;
				// }

				// log.Println(median)
			}
			count++
		}
	}

	log.Println("Done calculating features from images")

	f.AttemptUpdateText("Done calculating features")
	f.AttemptUpdateProBarValue(f.DatabaseHandler.CountRecords(f.featTableName))
	f.Finished()
}
